extra/utilities/blake3.py

```diff
@@ -0,0 +1,108 @@
+          states[b] = rotr(states[b] ^ states[c], 12 if m is mx else 7).cast(dtypes.uint32)
+      if i < 6: data = data[self.PERMUTATIONS]
+    return (states[:8] ^ states[8:]).cat(chain_vals[:8] ^ states[8:])
+
+  @TinyJit
+  def init_chain_vals(self, data: Tensor, info: Tensor) -> Tuple[Tensor, Tensor]:
+    chain_vals = self.IV.reshape(1, 8, 1).expand(16, 8, info.shape[-1]).contiguous()
+    counts = Tensor.arange(0, data.shape[-1], dtype=dtypes.uint32).reshape(-1, 1).expand(-1, 16).reshape(-1, 16, 1).permute(1, 2, 0)
+    counts = counts.cat(Tensor.zeros(chain_vals.shape[0], 1, chain_vals.shape[-1], dtype=dtypes.uint32), dim=1)
+    lengths = (info == self.DEFAULT_LEN).where(64, info)
+    flags = Tensor.zeros((16, 1, info.shape[-1]), dtype=dtypes.uint32).contiguous()
+    flags[-1, 0] = flags[-1, 0] + 2 # chunk end flag
+    flags = (flags + 2 * ((flags != 2) * (info < self.DEFAULT_LEN))) # chunk end flag for partial final chunk
+    flags[0] = flags[0] + 1 # chunk start flag
+    flags = (flags + (8 * (((info < self.PAD).sum() <= 16) * (info < self.DEFAULT_LEN)))).cast(dtypes.uint32) # root flag
+    states = (chain_vals.cat(chain_vals[:, :4], counts, lengths, flags, dim=1) * (info < self.PAD).cast(dtypes.uint32))
+    for i in range(16):
+      next_state = states[i] if i == 0 else states[i-1, :8].cat(states[i, 8:])
+      states[i] = self.compress_blocks(next_state, data[i], chain_vals[i])
+    states = states * (info < self.PAD)
+    end_block = (states * (info < self.DEFAULT_LEN)).sum(0)
+    return (states[-1, :] | end_block)[:8].realize()
+
+  @classmethod # JIT doesn't like making n_tree_steps for the loop a Variable, this is a workaround
+  def create_jitted_tree_hash(cls, n_tree_steps: int) -> Tensor:
+    def tree_hash(self, chain_vals: Tensor) -> Tensor:
+      for _ in range(n_tree_steps):
+        stacked = chain_vals.transpose().reshape(-1, 16).transpose().reshape(2, 8, -1)
+        stacked_mask = stacked.any(1)
+        final_step = chain_vals[0, :3].prod().cast(dtypes.bool).neg()
+        pair_mask, remainder_mask = (stacked_mask[0] * stacked_mask[1]), (stacked_mask[0] ^ stacked_mask[1])
+        paired, remainder = (stacked * pair_mask).reshape(16, -1), (stacked * remainder_mask).reshape(16, -1)[:8]
+        flags = final_step.where(12, Tensor.full((1, paired.shape[-1]), 4, dtype=dtypes.uint32))
+        iv = self.IV.reshape(8, 1).expand(8, paired.shape[-1])
+        counts = Tensor.zeros((2, paired.shape[-1]), dtype=dtypes.uint32)
+        lengths = Tensor.full((1, paired.shape[-1]), 64, dtype=dtypes.uint32)
+        states = iv.cat(iv[:4], counts, lengths, flags, dim=0)
+        chain_vals = ((self.compress_blocks(states, paired, iv) * pair_mask)[:8] + remainder).realize()
+        chain_vals = chain_vals.pad((None, (0, chain_vals.shape[1])))
+      return chain_vals.realize()
+    return TinyJit(tree_hash)
+
+  def tensor_to_blake_input(self, tensor: Tensor, padded_input_size: int) -> Tuple[Tensor, Tensor, Variable]:
```

Comment:
```suggestion
  def tensor_to_blake_input(self, tensor: Tensor, padded_input_size: int) -> Tuple[Tensor, Tensor, UOp]:
```
Looks like you tried using `Variable`(`UOp`) from tinygrad.ops, but here it’s actually `UOp.variable`.

Reply:
Ah, thanks for pointing that out!